Lib/test/test_sqlite3/test_dbapi.py

```diff
@@ -910,14 +935,15 @@ def test_cursor_executescript_with_surrogates(self):
                 insert into a(s) values ('\ud8ff');
                 """)
 
-    @unittest.skipUnless(sys.maxsize > 2**32, 'requires 64bit platform')
-    @bigmemtest(size=2**31, memuse=3, dry_run=False)
-    def test_cursor_executescript_too_large_script(self, maxsize):
-        con = sqlite.connect(":memory:")
-        cur = con.cursor()
-        for size in 2**31-1, 2**31:
-            with self.assertRaises(sqlite.DataError):
-                cur.executescript("create table a(s);".ljust(size))
+    def test_cursor_executescript_too_large_script(self):
+        msg = "query string is too large"
+        with memory_database() as cx, cx_limit(cx) as lim:
+            for sz in lim, lim+1:
```

Comment:
Seems there is no difference between `lim` and `lim+1`. We need to test both cases which raise and not raise an exception (i.e. for `lim-1` and `lim`).

What will happen if disable checks for `size >= max_length` in the C code? I expect the result be the same if size <= INT_MAX. Can you confirm?

Reply:
It would depend on the compile-time options `SQLITE_MAX_LENGTH` and `SQLITE_MAX_SQL_LENGTH`, and the current connection limits `SQLITE_LIMIT_LENGTH` and `SQLITE_LIMIT_SQL_LENGTH`. I wonder if our extra check is needed at all.

In any case, it will be good to test `lim-1`. I don't think there's a need to test `lim+1`. I'll ping you when I've done the needed changes. Thanks for reviewing!

Reply:
Ok, this is interesting, @serhiy-storchaka: I removed the `SQLITE_LIMIT_LENGTH` checks from the source code. After re-reading the SQLite documentation, it seems like we are checking the wrong limit; `SQLITE_LIMIT_LENGTH` has to do with string values. `SQLITE_LIMIT_SQL_LENGTH`, OTOH, limits the query length. Applying the patch below makes for correct behaviour:

```diff
diff --git a/Lib/test/test_sqlite3/test_dbapi.py b/Lib/test/test_sqlite3/test_dbapi.py
index c36f23f66d..d96075cda7 100644
--- a/Lib/test/test_sqlite3/test_dbapi.py
+++ b/Lib/test/test_sqlite3/test_dbapi.py
@@ -55,7 +55,7 @@ def memory_database():
 
 # Temporarily limit a database connection parameter
 @contextlib.contextmanager
-def cx_limit(cx, category=sqlite.SQLITE_LIMIT_LENGTH, limit=128):
+def cx_limit(cx, category=sqlite.SQLITE_LIMIT_SQL_LENGTH, limit=128):
     try:
         _prev = cx.setlimit(category, limit)
         yield limit
@@ -936,11 +936,11 @@ def test_cursor_executescript_with_surrogates(self):
                 """)
 
     def test_cursor_executescript_too_large_script(self):
-        msg = "query string is too large"
+        msg = "string or blob too big"
         with memory_database() as cx, cx_limit(cx) as lim:
-            cx.executescript("select 'almost too large'".ljust(lim-1))
+            cx.executescript("select 'almost too large'".ljust(lim))
             with self.assertRaisesRegex(sqlite.DataError, msg):
-                cx.executescript("select 'too large'".ljust(lim))
+                cx.executescript("select 'too large'".ljust(lim+1))
 
     def test_cursor_executescript_tx_control(self):
         con = sqlite.connect(":memory:")
diff --git a/Lib/test/test_sqlite3/test_regression.py b/Lib/test/test_sqlite3/test_regression.py
index 158f4cf86f..86742713e9 100644
--- a/Lib/test/test_sqlite3/test_regression.py
+++ b/Lib/test/test_sqlite3/test_regression.py
@@ -358,15 +358,15 @@ def test_surrogates(self):
         self.assertRaises(UnicodeEncodeError, cur.execute, "select '\udcff'")
 
     def test_large_sql(self):
-        msg = "query string is too large"
+        msg = "string or blob too big"
         with memory_database() as cx, cx_limit(cx) as lim:
             cu = cx.cursor()
 
-            cx("select 1".ljust(lim-1))
+            cx("select 1".ljust(lim))
             # use a different SQL statement; don't reuse from the LRU cache
-            cu.execute("select 2".ljust(lim-1))
+            cu.execute("select 2".ljust(lim))
 
-            sql = "select 3".ljust(lim)
+            sql = "select 3".ljust(lim+1)
             self.assertRaisesRegex(sqlite.DataError, msg, cx, sql)
             self.assertRaisesRegex(sqlite.DataError, msg, cu.execute, sql)
 
diff --git a/Modules/_sqlite/cursor.c b/Modules/_sqlite/cursor.c
index 1d7c0b46a6..9b0d62b5fe 100644
--- a/Modules/_sqlite/cursor.c
+++ b/Modules/_sqlite/cursor.c
@@ -728,13 +728,6 @@ pysqlite_cursor_executescript_impl(pysqlite_Cursor *self,
     self->reset = 0;
 
     size_t sql_len = strlen(sql_script);
-    int max_length = sqlite3_limit(self->connection->db,
-                                   SQLITE_LIMIT_LENGTH, -1);
-    if (sql_len >= (unsigned)max_length) {
-        PyErr_SetString(self->connection->DataError,
-                        "query string is too large");
-        return NULL;
-    }
 
     // Commit if needed
     sqlite3 *db = self->connection->db;
diff --git a/Modules/_sqlite/statement.c b/Modules/_sqlite/statement.c
index b20c91da31..898d5f106f 100644
--- a/Modules/_sqlite/statement.c
+++ b/Modules/_sqlite/statement.c
@@ -60,12 +60,6 @@ pysqlite_statement_create(pysqlite_Connection *connection, PyObject *sql)
     }
 
     sqlite3 *db = connection->db;
-    int max_length = sqlite3_limit(db, SQLITE_LIMIT_LENGTH, -1);
-    if (size >= max_length) {
-        PyErr_SetString(connection->DataError,
-                        "query string is too large");
-        return NULL;
-    }
     if (strlen(sql_cstr) != (size_t)size) {
         PyErr_SetString(PyExc_ValueError,
                         "the query contains a null character");
```

Reply:
(We only need to fix this in `main`)

Reply:
We need an extra check to prevent an integer overflow (if size > INT_MAX). For size <= INT_MAX our check should be equal to the SQLite check.

I think that changes in the limit category, error message and inclusivity of the limit should be backported. Is it what you meant?